src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,569 @@
+  return handle.enter(write_pipeline->device_submission
+  ).then([write_fut = std::move(write_fut)]() mutable {
+    return std::move(write_fut
+    ).handle_error(
+      write_ertr::pass_further{},
+      crimson::ct_error::assert_all{
+        "Invalid error in CircularBoundedJournal::append_record"
+      }
+    );
+  }).safe_then([this, &handle] {
+    return handle.enter(write_pipeline->finalize);
+  }).safe_then([this, target,
+    length=to_write.length(),
+    &write_result,
+    &r_size,
+    FNAME] {
+    DEBUG(
+      "append_record: commit target {} used_size {} written length {}",
+      target, get_used_size(), length);
+
+    set_last_committed_record_base(target);
+    set_used_size(get_used_size() + length);
+    paddr_t paddr = convert_abs_addr_to_paddr(
+      target + r_size.get_mdlength(),
+      header.device_id);
+    auto submit_result = record_locator_t{
+      paddr,
+      write_result
+    };
+    return submit_result;
+  });
+}
+
+CircularBoundedJournal::write_ertr::future<> CircularBoundedJournal::device_write_bl(
+    rbm_abs_addr offset, bufferlist &bl)
+{
+  LOG_PREFIX(CircularBoundedJournal::device_write_bl);
+  auto length = bl.length();
+  if (offset + length > header.end) {
+    return crimson::ct_error::erange::make();
+  }
+  bl.rebuild_aligned(get_block_size());
+  DEBUG(
+    "overwrite in CircularBoundedJournal, offset {}, length {}",
+    offset,
+    length);
+  auto write_length = length < get_block_size() ? get_block_size() : length;
+  auto bptr = bufferptr(ceph::buffer::create_page_aligned(write_length));
+  auto iter = bl.cbegin();
+  iter.copy(bl.length(), bptr.c_str());
+  return device->write(offset, bptr
+  ).handle_error(
+    write_ertr::pass_further{},
+    crimson::ct_error::assert_all{ "Invalid error device->write" }
+  ).safe_then([] {
+    return write_ertr::now();
+  });
+}
+
+CircularBoundedJournal::read_super_ret
+CircularBoundedJournal::read_super(rbm_abs_addr start)
+{
+  LOG_PREFIX(CircularBoundedJournal::read_super);
+  auto bptr = bufferptr(ceph::buffer::create_page_aligned(
+			device->get_block_size()));
+  return device->read(start, bptr
+  ).safe_then([start, bptr, FNAME]() mutable
+    -> read_super_ret {
+    DEBUG("read_super: reading {}", start);
+    bufferlist bl;
+    bl.append(bptr);
+    auto bp = bl.cbegin();
+    cbj_header_t cbj_header;
+    try {
+      decode(cbj_header, bp);
+    } catch (ceph::buffer::error &e) {
+      ERROR("read_super: unable to read super block");
+      return crimson::ct_error::enoent::make();
+    }
+    return read_super_ret(
+      read_super_ertr::ready_future_marker{},
+      std::make_pair(cbj_header, bl)
+    );
+  });
+}
+
+Journal::replay_ret CircularBoundedJournal::replay(
+    delta_handler_t &&delta_handler)
+{
+  /*
+   * read records from last applied record prior to written_to, and replay
+   */
+  LOG_PREFIX(CircularBoundedJournal::replay);
+  if (get_last_committed_record_base() == get_applied_to()) {
```

Comment:
I'm probably misunderstanding the semantics here, but this seems wrong.  replay() gets invoked prior to open_for_write() and I don't see anywhere else where the header gets read.

Even if the header does get read prior to this point, I think there's still a problem.  I see set_last_committed_record_base() being called at the end of submit_record prior to completion -- that seems right.  However, set_last_committed_record_base() updates the in-memory copy of header without performing a write-out.  If close() isn't called (perhaps because the node in question was power-cycled), rereading the header will result in a badly out of date header with a value of last_committed_record_base which is badly out of date.

For that reason, I'm somewhat confused by the purpose of cbj_header_t::last_committed_record_base.  Since you can't rely on a clean shutdown, its value must necessarily be out of date unless you plan on rewriting the header for every record.  Is it meant as a way to detect broken/corrupted devices?

Reply:
I would expect that replay() would find written_to and last_committed_record_base from the last readable record rather than relying on the value in the header.

Reply:
> I'm probably misunderstanding the semantics here, but this seems wrong. replay() gets invoked prior to open_for_write() and I don't see anywhere else where the header gets read.

Added.


> Even if the header does get read prior to this point, I think there's still a problem. I see set_last_committed_record_base() being called at the end of submit_record prior to completion -- that seems right. However, set_last_committed_record_base() updates the in-memory copy of header without performing a write-out. If close() isn't called (perhaps because the node in question was power-cycled), rereading the header will result in a badly out of date header with a value of last_committed_record_base which is badly out of date.

Right. To address the issue you mentioned, I added a commit (https://github.com/ceph/ceph/pull/44233/commits/77e3799a3c94c7bd890a111021230ce4bd9c0853). The intention here is that CBJournal makes applied_to (the tail of journal) persistent whenever applying---apply the content of records in CBjournal to RBM---occurs, not every writes (submit_record). Since replay() will be started from appliied_to, it can recover last_committed_records_base and written_to correctly if we maintain applied_to safely. Note that replay() try to read record sequentially until it finds invalid records. Make sense?
In addition, I have plan to implement applied_to update---storing applied_to persistent---with cleaner changes (future work item).